tests/conftest.py

```diff
@@ -126,18 +168,70 @@ def cert_encrypted_private_key_file(example_cert):
         yield tmp
 
 
+class TestServer(Server):
+    # This customized uvicorn server encapsulates the server restart logic.
+    # We wrap shutdown()/startup() so that the server, which is running on asyncio in
+    # the main thread, can be restarted without closing the connection, even from
+    # another thread and/or async environment.
+
+    def __init__(self, config):
+        super().__init__(config)
+        self._restart_channel = asyncio.Queue()
+
+    async def restart(self) -> None:
+        # Request a restart, and wait for it to complete.
+        await self._restart_channel.put(None)
+        await self._restart_channel.get()
```

Comment:
This is completely broken: *obviously* this returns immediately since we put a `None` in the queue and remove it immediately after. 🤦‍♂ 

I tried an event-based implementation instead, but we're mixing event loops and asyncio doesn't agree.

I'm starting to wonder if we could just turn towards hypercorn for the background server… Would definitely resolve a lot of difficulties here to be able to run everything in the main thread.

Reply:
What's *super weird* though is that the tests *passed*, as if restarting the server didn't make any difference. 😬 (And indeed, commenting the `restart_server()` bit in the connection pool tests has no effect: the tests pass.)